asm-to-pil/src/vm_to_constrained.rs

```diff
@@ -454,6 +456,86 @@ impl<T: FieldElement> ASMPILConverter<T> {
+                rhs
+                .inputs
+                .iter_mut()
+                .chain(rhs.outputs.iter_mut())
+                .for_each(|p| {
+                    assert!(p.ty.is_none());
+                    let reg = self.registers.get(&p.name).expect("rhs params must be registers");
+                    match reg.ty {
+                        RegisterTy::Assignment => {
+                            let mut lhs_params = lhs.inputs.iter().chain(lhs.outputs.iter());
+                            assert!(lhs_params.any(|p_lhs| p_lhs.name == p.name),
+                                    "assignment register on rhs must be present on lhs params");
+                        }
+                        RegisterTy::Write => {
+                            // the linker needs this information to build the
+                            // plookup when a write register is used as output.
+                            // TODO: less hackish way of passing this info? proper type for param.ty?
+                            p.ty = Some("write".to_string());
+                        },
+                        _ => panic!("rhs of instruction/link declaration must only use write or assignemtn registers"),
+                    }
+                });
+
+                // all lhs params must be used on rhs
+                lhs.inputs.iter().chain(lhs.outputs.iter()).for_each(|p| {
+                    assert!(
+                        rhs.inputs
+                            .iter()
+                            .chain(rhs.outputs.iter())
+                            .any(|rhs_p| rhs_p.name == p.name),
+                        "all lhs registers must be used on rhs"
+                    )
+                });
+
+                // can't repeat registers on output of rhs
+                let uniq: BTreeSet<_> = rhs.outputs.iter().map(|p| &p.name).collect();
+                assert_eq!(
+                    rhs.outputs.len(),
+                    uniq.len(),
+                    "rhs of instruction/link declaration can't have repeated output registers"
+                );
+            }
+
+            // if a write register is used as output in a instr/link mapping, we
+            // must induce a tautology in the update clause when that
+            // instruction is active, to allow the operation plookup to match.
+            for p in rhs.outputs.iter() {
+                let reg = self
+                    .registers
+                    .get_mut(&p.name)
+                    .expect("rhs params must be registers");
+                if reg.ty.is_write() {
+                    let Expression::Reference(ref poly) = l.flag else {
+                        panic!();
+                    };
+                    let flag = poly.try_to_identifier().unwrap();
+                    let reference = direct_reference(flag);
+                    let value = next_reference(&p.name);
+                    reg.conditioned_updates.push((reference, value));
```

Comment:
what does this do?

Reply:
when the instruction is active, `C' = C'`, when it is not active `C'` behaves as usual (depends on everything else)